src/engine/V3Eval.v3

```diff
@@ -1090,32 +1090,30 @@ def do_v128_bitmask<T>(a: (u64, u64), bits: byte, bits_log2:byte) -> i32 {
 // T: type to be converted to, having the same width as source type
 // Tw: wider type of T, i.e., the extended type
 // Uw: corresponding unsignced type of Tw
-// S: type of v128.{low/high}, which is always u64, but need to be explicit put here for type inference
-def do_v128_half_extend<T, Tw, Uw>(a: (u64, u64), low_half: bool, dest_bits: byte,
-convert: u64 -> T, extend: (T) -> Tw, convert_to_u: (Tw) -> Uw) -> (u64, u64) {
+def do_v128_half_extend<T, Tw, Uw>(a: (u64, u64), low_half: bool, dest_bits: byte, 
+	convert: u64 -> T, extend: (T) -> Tw, convert_to_u: (Tw) -> Uw) -> (u64, u64) {
 	var src = if(low_half, a.0, a.1);	// Choose low or high half of source lanes.
 	var src_high: u32 = u32.view(src >> 32);
 	var src_low: u32 = u32.view(src);
-	// Extend low and high lanes separately.
 	var dst_low = do_32_ext_64<T, Tw, Uw>(src_low, dest_bits, convert, extend, convert_to_u);
 	var dst_high = do_32_ext_64<T, Tw, Uw>(src_high, dest_bits, convert, extend, convert_to_u);
 	return (dst_low, dst_high);
 }
-def do_32_ext_64<T, Tw, Uw>(src: u32, dest_bits: byte,
-convert: u64 -> T, extend: (T) -> Tw, convert_to_u: (Tw) -> Uw) -> u64 {
+def do_32_ext_64<T, Tw, Uw>(src: u32, dest_bits: byte, 
+	convert: u64 -> T, extend: (T) -> Tw, convert_to_u: (Tw) -> Uw) -> u64 {
 	var res: u64 = 0u;
 	var shift = byte.view(dest_bits >> 1);	// Compute source shift amount, i.e., the width of the original lane.
-	var iters = byte.view(64 / dest_bits);	// Compute number of iterations. Can be deduced from "shift", put here for clarity.
-	for (i < iters) {
+	var dst_shift = shift << 1;
+	for (i = 0; i < 64; i += dst_shift) {
 		var val = extend(convert(u64.view(src)));
 		src >>= shift;
-		var res_shift = byte.view(shift * 2 * i);	// Compute shift amount for result.
```

Comment:
Can also improve this to a `u6.view`, which will save a another shift check underneath.

Reply:
No prob! Missed this one. Will push a change soon.

Reply:
Updated:)